chirp/drivers/kenwood_live.py

```diff
@@ -84,7 +84,7 @@ def command(ser, cmd, *args):
 
     result = ""
     while not result.endswith(LAST_DELIMITER[0]):
-        result += ser.read(COMMAND_RESP_BUFSIZE).decode()
+        result += ser.read(COMMAND_RESP_BUFSIZE).decode('ascii', 'replace')
```

Comment:
Why? The kenwood live protocol is entirely ASCII. I think anywhere you do this `decode('ascii', 'replace')` is just a hack to convert a bytestring into a unicode string, but with silent failures for unexpected values. It's much better to avoid this whenever possible.

Reply:
The decode() function by default is very strict. During probing the serial port for the correct speed some unknown character ends up in decode causing an UnicodeError. This stop the probing before reaching the correct radio speed.

![Screenshot from 2022-07-04 23-46-08](https://user-images.githubusercontent.com/5526083/196882417-21da26c5-8db0-4a7d-aa82-7c6e893af3d1.png)

``[2022-10-20 10:15:07,216] chirp.drivers.kenwood_live - INFO: Trying ID at baud 57600 with delimiter "('\r', ' ')"

[2022-10-20 10:15:07,327] chirp.drivers.kenwood_live - DEBUG: PC->RADIO: ID
[2022-10-20 10:15:07,828] chirp.drivers.kenwood_live - ERROR: Timeout waiting for data
[2022-10-20 10:15:07,828] chirp.drivers.kenwood_live - ERROR: Giving up
[2022-10-20 10:15:07,828] chirp.drivers.kenwood_live - INFO: Trying ID at baud 38400 with delimiter "('\r', ' ')"
[2022-10-20 10:15:07,939] chirp.drivers.kenwood_live - DEBUG: PC->RADIO: ID
[2022-10-20 10:15:08,440] chirp.drivers.kenwood_live - ERROR: Timeout waiting for data
[2022-10-20 10:15:08,440] chirp.drivers.kenwood_live - ERROR: Giving up
[2022-10-20 10:15:08,440] chirp.drivers.kenwood_live - INFO: Trying ID at baud 19200 with delimiter "('\r', ' ')"
[2022-10-20 10:15:08,551] chirp.drivers.kenwood_live - DEBUG: PC->RADIO: ID
``
I think that relaxing the decode() is ok since the protocol is ASCII as you said.

Reply:
Right, but this is exactly the point. The error indicates that we're getting something we don't expect. Making `command()` ignore all of these just means we won't be able to tell when something is going wrong, we'll just get mangled (or no) data back from the radio. If our cable is broken or noisy, blindly filtering out the non-ascii bits could cause us to do something bad to the radio, misinterpret a response, etc.

`command()` expects to converse with the radio, so it should not ignore those errors. However, the `get_id()` routine is probing for a radio, trying different rates, etc, so the situation is different. So here:

https://github.com/kk7ds/chirp/blob/py3/chirp/drivers/kenwood_live.py#L124

We're using `command()` but would not be surprised to get a protocol-violating error due to the fact that we don't know if we're speaking the correct baud rate. So you should `try..except UnicodeDecodeError` around the `command()` invocation and if we get one, don't even bother inspecting `resp` as you know it's not going to be what we want. That way, all the other users of `command()` will abort the conversation if we get unexpected non-ASCII bytes, which is what should happen.